README
* Introduction

This is a modified version of Douglas Laing mix package. The original
version is available from http://home.telkomsa.net/dlaing/.

Please see COPYING for information about copying conditions.

The file NEWS contains a list of user-visible changes to the original
package.

* Installation

The file INSTALL contains generic build instructions. In short, run

 $ ./configure
 $ make
 $ make install

So far, there is only one package-specific configuration option: 
--without-readline.  Use it if you wish to disable readline support in 
MIX terminal.

* The Assembler  -  mixal

Mixal reads MIXAL source code from the file named as an argument (or
from stdin if no filename is given) and assembles it. If assembly
succeeds, mixal ewrites the output load deck to stdout. The load deck
consists of a two card loading routine followed by instruction cards
and a transfer card. (See TAOCP 1.3.1 exercise 26.) Any error messages
generated are written to stderr.

Mixal exits with zero exit code if assembly succeeded, otherwise its
exit code is 1. 

Mixal accepts several command line options that alter its
behavior. The option --list (-l) causes the assembled instructions to be
written to a file. The file name is obtained by replacing the input
file suffix with .lst, e.g. listing file for program `easter.mix' will
be named `easter.lst'. If the source file name does not have any
suffix, `.lst' is appended to it. Otherwise, if the source is taken from
the standard input, the listing file is named `mixal.lst'

The listing file ends with a dump of the symbol table used during
assembly.

The option --xref (-x) instructs mixal to produce a cross-reference
listing of the assembled program. If --list is also given, this listing
is included into the last part of the listing file. Otherwise, it is
output to stderr.

The option --raw-output (-r) instructs mixal to output the `raw code',
i.e. to not convert it to usual MIX card format and to not output the
usual card loading routine and transfer card. This option is useful
for writing card loaders.

The option --force (-f) forces mixal to produce output files (card
deck and listing file, if required) even if there were errors during
the assembly.

The options --debug-gram and --debug-lex enable debugging output for
the parser and lexical analyzer, correspondingly. The option -d,
provided for backward compatibility with the version 1.2, allows
to set both debugging features at once. It accepts an argument
consisting of at most two letters. The letter `y' enables tracing of
input file grammar analyzer, while the letter `l' (ell) enables
tracing of input file lexical analyzer. Thus, the full debugging
output can be obtained by running `mixal -d yl'.

There are also two informational options:  --version (-V) which
reports the version number and --help (-h) which displays a short
usage summary. 

* The simulator -  mixsim

When run without command line options, mixsim reads a load deck from
sysin, loads and executes it, writing any error messages to stderr. On
completion a dump of the machine state is written to stderr. If mixsim
is interrupted during execution, the machine state will be dumped to
stderr.

By default the card reader is connected to sysin, the printer to
sysout and the card punch to stderr. The tape units use files tape0 to
tape7 and the disk units use files disk8 to disk15 in the current
directory.

These defaults can be altered by giving the --assign-device (-a)
command line option. The option takes a single argument in the form

  DEV=NAME

Its effect is to bind MIX device #DEV to UNIX file NAME (or, assign
the UNIX file NAME to MIX device #DEV, hence the option name
`--assign-device')

The disks are always opened for update without truncation, so that old
data is not destroyed until it is overwritten. Note that big disk
files will never shrink unless they are deleted.

The machine is shipped without a papertape unit and without the
floating point feature.

* The terminal and the debugger

When run with the --terminal (-t) command line option, mixsim starts
up in MIX terminal mode. This mode provides a set of commands for
running and debugging MIX programs. The commands are case-insensitive,
each command has a full and contracted form. Furthermore, if MIX is
compiled with readline support, any unambiguous abbreviation of the
command name is also allowed.

As a shortcut, hitting newline in terminal mode repeats the recent
command.

The list of available terminal commands is given below. For the sake
of consistency the commands are given in upper case. The full form is
given after two *'s, the contracted form is listed on the next line.

** ADDRESS
   AD

This command takes as its argument a breakpoint manipulation command
(see DELETE, IGNORE, ENABLE, DISABLE, PASSCOUNT below) with arguments
appropriate for the associated command.  However, the first argument
of the corresponding command is treated as MIX address, instead of a
breakpoint number.

** ASGN <DEV> <FILE>
   A <DEV> <FILE>

Assign or bind a file to MIX device. For example, the command:

   ASGN 16 easter.mix

binds MIX card reader (device 16) to the UNIX file `easter.mix'. 

** BREAK <ADDR>
   B <ADDR>

Set breakpoint on address <ADDR>

** BREAK TEMP <ADDR>
   BT <ADDR>
   TB <ADDR>

Set a temporary breakpoint on address <ADDR>. Such a breakpoint is
deleted after the first crossing.

This command is equivalent to:

BREAK <ADDR>
PASSCOUNT <N> 1

** CLEAR <N>
   CB <N>
   DELETE <N>

Delete the breakpoint #<N>.
See also ADDRESS.

* CONT
  CO

Continue after encountering a breakpoint.

** DISABLE <N>

Disable breakpoint #<N>.
See also ENABLE and ADDRESS.

** DISASSEMBLE [<BEG> [<END>]]

See UNASM below

** DUMP
   DU

Dump entire MIX machine state

** DUMP REGISTERS
   DR

Dump MIX registers

** DUMP MEMORY [<BEG> [<END>]]
   DM [<BEG> [<END>]]

Dump the contents of MIX memory. If no arguments are given, the entire
memory range (0-4000) is printed. Otherwise, <BEG> and optional <END>
specify start and end addresses of the memory range to dump.

** ENABLE <N>

Enable breakpoint #<N>.
See also DISABLE and ADDRESS.

** IGNORE <N> <COUNT>

Ignore next <COUNT> crossings of breakpoint #<N>.
See also ADDRESS.

** GO
   G

Emulates GO button. Typing GO starts execution of the current program.

** LIST BREAK
   LB

List all breakpoints

* LIST IO [<DEV>]
  LI [<DEV>]

List MIX I/O devices and their states.

When given an argument, the command will display the status of the
given device.

** PASSCOUNT <N> <COUNT>

Delete (clear) breakpoint <N> after <COUNT> crossings.
See also ADDRESS.

** NEXT
   N

Execute the next MIX instruction.

** RUN
   R

Same as GO.   

** STEP
   S

Execute the next MIX instruction. If it is a function call, step into
the function being called.

** UNASM [<BEG> [<END>]]
   U [<BEG> [<END>]]

Disassemble the given memory range. <BEG> and <END> have the same
meaning as for DU.

* QUIT
  Q

Quit the terminal. Typing ^D (end-of-file symbol) has the same effect.

* Debugging tips.

1. The usual terminal session starts with a command

ASGN 16 <DECK>

(where <DECK> is the name of UNIX file containing MIX card deck).

2. Set the breakpoints using BREAK command.

3. Finally, run GO

4. Upon encountering a breakpoint, the terminal stops and displays the
instruction to be executed next. Do whatever it is necessary to do
(e.g. examine the memory or registers, set new breakpoints or clear
old ones, etc) and type CONT to resume execution of the program.

5. You may restart the program after any breakpoint by typing GO again.

6. To exit the terminal, type end-of-file (Ctrl-D) or QUIT.


Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


NEWS
MIX NEWS -- history of user-visible changes. 2007-09-23
Copyright (C) 1998,1999,2000,2005,2007 Douglas Laing, Sergey Poznyakoff

See the end of file for copying conditions.

Please send bug reports to <gray@gnu.org.ua>.


Version 2.0, 2007-09-23

* Licensed under GPLv3 or later.

* mixal

** Location of listing file

By default, mixal creates listing file in the current working directory.

* mixsim

** The .mixsim file
When started in terminal mode, the utility searches for file `.mixsim'
in the current working directory and reads it, if present. The file
may contain any mixsim commands.

** The set of terminal commands is improved to make it compatible with GDB.

** New command, PASSCOUNT
Set the number of crossings after whcih to clear the breakpoint.

** Removed optional second argument to BREAK.
The functionality of `BREAK <ADDR> <COUNT>' is now obtained by two
commands:

BREAK <ADDR>
PASSCOUNT <N> <COUNT>

** New command DELETE.
An alias to CLEAR.

** New command IGNORE
Sets the number of hits to ignore before enabling the breakpoint.

** New commands ENABLE and DISABLE
Enable or disable the breakpoint with the given number.

** New command RUN
It is an alias for GO.

** New command SOURCE
Execute a file of mixsim commands.

** New command ADDRESS.
This command takes as its argument a breakpoint manipulation command
(DELETE, IGNORE, ENABLE, DISABLE, PASSCOUNT) with arguments
appropriate for the associated command.  However, the first argument
of the corresponding command is treated as MIX address, instead of a
breakpoint number.

** New command INFO.
An alias to LIST.

** New command SHELL.
Executes a subordinate shell or a shell command. Abbreviated to !.

** NEXT and STEP take optional argument.
The argument, if given, specifies number of instructions to
execute. Default is 1.

** INFO BREAK takes an optional argument specifying breakpoint number to list.
If prefixed with ADDRESS, it treats the argument as a MIX location and
lists all breakpoints set on that location.


Version 1.2

* New terminal mode (mixsim -t) allows to debug MIXAL programs
* New option -r to mixal allows to produce 'raw' MIX programs. It is useful to
experiment with loader routines.
* Mixal can produce a cross-reference listing (mixal -x).
* Mixal prints a warning if an ALF constant is preceeded by more than two 
spaces and followed by a printable character.


* Previous entries by Douglas Laing

99/04/22  remove requirement for trailing blanks in ALF statements
99/04/13  version 1.02
99/04/13  gcc compilation of mixal generates many integer overflow messages
99/04/13  mixal uses locations instead of line numbers to order local symbols
98/08/19  version 1.01
98/08/19  mixal crashes under linux because of a null pointer


Copyright information:

Copyright (C) 1998, 1999, 2000 Douglas Laing
Copyright (C) 2000, 2005, 2007 Sergey Poznyakoff

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.


Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
eval: (add-hook 'write-file-hooks 'time-stamp)
time-stamp-start: "changes. "
time-stamp-format: "%:y-%02m-%02d"
time-stamp-end: "\n"
end:


yfmt/yfmt.c
/*2:*/
#line 46 "./yfmt.w"

#ifdef HAVE_CONFIG_H
# include <config.h> 
#endif
#include <stdio.h> 
#include <stdlib.h> 
#include <unistd.h> 
#include <stdarg.h> 
#include <errno.h> 
#include <assert.h> 
#ifdef HAVE_GETOPT_H
# include <getopt.h> 
#endif

/*:2*//*3:*/
#line 61 "./yfmt.w"

typedef struct{
char*file;
int line;
}Locus;

/*:3*//*4:*/
#line 71 "./yfmt.w"

Locus real_locus,effective_locus;

/*:4*//*6:*/
#line 80 "./yfmt.w"

FILE*infile,*outfile;

/*:6*//*9:*/
#line 102 "./yfmt.w"

void
output_effective_line()
{
fprintf(outfile,"#line %d \"%s\"\n",
effective_locus.line,effective_locus.file);
}

/*:9*//*10:*/
#line 111 "./yfmt.w"

void
output_chars(char*p)
{
fprintf(outfile,"%s",p);
}

/*:10*//*11:*/
#line 119 "./yfmt.w"

char*progname;

/*:11*//*13:*/
#line 131 "./yfmt.w"

void
error(char*fmt,...)
{
va_list ap;

fprintf(stderr,"%s: ",progname);
va_start(ap,fmt);
vfprintf(stderr,fmt,ap);
va_end(ap);
fprintf(stderr,"\n");
}

/*:13*//*14:*/
#line 145 "./yfmt.w"

void
usage_error(char*msg)
{
error(msg);
fprintf(stderr,"Try `%s -h' for more information.\n",progname);
exit(1);
}

/*:14*//*15:*/
#line 155 "./yfmt.w"

void
input_error(char*msg)
{
error("%s:%d: %s",effective_locus.file,effective_locus.line,msg);
}

/*:15*//*16:*/
#line 203 "./yfmt.w"

#define LIMBO 0
#define CHEAD 1
#define DEFNS 2
#define RULES 3
#define CBODY 4
#define CTAIL 5 \

#define EOL 1
#define LINE 2
#define YSEP 3
#define YLBRACE 4
#define YRBRACE 5
#define COMMENT 6 \
 \

#define MAXSAVECHARS 128 \


#line 204 "./yfmt.w"

int state;

/*:16*//*17:*/
#line 208 "./yfmt.w"

typedef void(*Procfun)(void);

/*:17*//*18:*/
#line 212 "./yfmt.w"

/*21:*/
#line 232 "./yfmt.w"

int token;
int lastchar= '\n';

/*:21*//*22:*/
#line 240 "./yfmt.w"

int return_line_info= 0;

/*:22*//*23:*/
#line 249 "./yfmt.w"

char pbuffer[MAXSAVECHARS];
int pindex= 0;

/*:23*//*24:*/
#line 254 "./yfmt.w"

int
nextchar()
{
if(pindex> 0)
lastchar= pbuffer[--pindex];
else{
lastchar= fgetc(infile);
if(lastchar=='\n'){
real_locus.line++;
effective_locus.line++;
}
}
return lastchar;
}

/*:24*//*25:*/
#line 272 "./yfmt.w"

char savebuffer[MAXSAVECHARS];
int saveindex;

/*:25*//*26:*/
#line 277 "./yfmt.w"

void
putback()
{
while(saveindex> 0)
pbuffer[pindex++]= savebuffer[--saveindex];
}

/*:26*//*27:*/
#line 286 "./yfmt.w"

void
putback_last()
{
pbuffer[pindex++]= lastchar;
}

/*:27*//*28:*/
#line 294 "./yfmt.w"

void
clearsave()
{
saveindex= 0;
}

/*:28*//*29:*/
#line 302 "./yfmt.w"

void
savechar()
{
if(saveindex==MAXSAVECHARS){
input_error("Save buffer overflow");
abort();
}
savebuffer[saveindex++]= lastchar;
}

/*:29*//*30:*/
#line 314 "./yfmt.w"

void
skipchars()
{
while(nextchar()!=EOF&&lastchar!='\n')
;
}

/*:30*//*31:*/
#line 324 "./yfmt.w"

void
save_effective_line()
{
saveindex= sprintf(savebuffer,"#line %d \"%s\"\n",
effective_locus.line,effective_locus.file);
if(saveindex==MAXSAVECHARS){
input_error("Save buffer overflow");
abort();
}
}

/*:31*//*32:*/
#line 337 "./yfmt.w"

void
flushsave()
{
if(saveindex){
fwrite(savebuffer,saveindex,1,outfile);
clearsave();
}
}

/*:32*//*33:*/
#line 349 "./yfmt.w"

void
copychars()
{
if(token==EOL)
fputc('\n',outfile);
else{
flushsave();
while(nextchar()!=EOF&&lastchar!='\n')
fputc(lastchar,outfile);
if(lastchar=='\n')
fputc(lastchar,outfile);
}
}

/*:33*//*34:*/
#line 365 "./yfmt.w"

int
get_token()
{
static char*linestr= "line";

if(lastchar!='\n')
skipchars();

if(nextchar()==EOF)
return EOF;

if(lastchar=='\n')
return EOL;
else if(lastchar=='#'){
int n;
char*p;

savechar();
while(nextchar()&&isspace(lastchar))
;

for(p= linestr;*p;p++,nextchar()){
savechar();
if(lastchar!=*p)
return LINE;
}

while(nextchar()&&isspace(lastchar))
;

clearsave();
if(!isdigit(lastchar))
input_error("Invalid #line directive");
else{
for(n= 0;isdigit(lastchar);nextchar())
n= n*10+lastchar-'0';

effective_locus.line= n-1;

while((lastchar!='\n'&&isspace(lastchar))&&nextchar())
;

if(lastchar=='"'){

while(nextchar()&&lastchar!='"')
savechar();

effective_locus.file= malloc(saveindex+1);
assert(effective_locus.file!=NULL);
memcpy(effective_locus.file,savebuffer,saveindex);
effective_locus.file[saveindex]= 0;

clearsave();
}
if(return_line_info){
save_effective_line();
return LINE;
}
}
}else if(lastchar=='%'){
savechar();
nextchar();
savechar();
switch(lastchar){
case'{':
return YLBRACE;

case'}':
return YRBRACE;

case'%':
return YSEP;

default:
return LINE;
}
}else if(lastchar=='/'){
savechar();
if(nextchar()=='*'){
do{
do
savechar();
while(nextchar()&&lastchar!='*');
savechar();
}while(nextchar()&&lastchar!='/');
savechar();
return COMMENT;
}else
return LINE;
}else{
savechar();
return LINE;
}
return get_token();
}

/*:34*//*35:*/
#line 463 "./yfmt.w"

int
gettkn()
{
return token= get_token();
}

/*:35*/
#line 213 "./yfmt.w"

/*36:*/
#line 471 "./yfmt.w"

void
proc_limbo()
{
return_line_info= 0;
while(state==LIMBO&&gettkn()> 0){
if(token==YLBRACE)
state= CHEAD;
else if(token==YSEP)
state= RULES;
else if(saveindex> 0&&savebuffer[0]=='%')
copychars();
else
clearsave();
}
}

/*:36*//*37:*/
#line 489 "./yfmt.w"

void
proc_chead()
{
return_line_info= 1;
copychars();
output_effective_line();
while(state==CHEAD&&gettkn()> 0){
copychars();
if(token==YRBRACE)
state= DEFNS;
}
}

/*:37*//*38:*/
#line 504 "./yfmt.w"

void
proc_defns()
{
return_line_info= 0;
while(state==DEFNS&&gettkn()> 0){
if(token==COMMENT)
fputc('\t',outfile);
copychars();
if(token==YSEP)
state= RULES;
}
}

/*:38*//*40:*/
#line 544 "./yfmt.w"

void
proc_rules()
{
int ws= 0;

return_line_info= 0;

while(state==RULES&&gettkn()> 0){
if(token==YSEP)
state= CTAIL;
else if(token==EOL)
;
else if(token==COMMENT){
fputc('\t',outfile);
copychars();
}else if(token!=EOL){
int quote= 0;
int escape= 0;

putback();
do{
nextchar();
if(escape)
escape= 0;
else if(quote){
if(lastchar==quote)
quote= 0;
}else if(lastchar=='\\')
escape= 1;
else if(lastchar=='"')
quote= '"';
else if(lastchar=='[')
quote= ']';
else if(lastchar=='\'')
quote= '\'';
else if(lastchar==' '||lastchar=='\t')
ws= 1;
else if(lastchar=='{'){
if(ws){
state= CBODY;
break;
}
}
savechar();
}while(lastchar!='\n');
ws++;
flushsave();
}
}
}

/*:40*//*41:*/
#line 597 "./yfmt.w"

void
proc_cbody()
{
int nesting= 0;
int quote= 0;

return_line_info= 1;

output_chars("\t{");

for(nextchar();lastchar==' '||lastchar=='\t';nextchar())
fputc(lastchar,outfile);

if(lastchar!='\n'){
fputc('\n',outfile);
putback_last();
}else{
fputc(lastchar,outfile);
if(gettkn()<0)
return;
putback();
}

output_effective_line();
while(1){
int escape= 0;

if(!strchr(" \t#",lastchar)&&!quote)
fputc('\t',outfile);
do{
nextchar();
if(escape)
escape= 0;
else if(quote){
if(lastchar==quote)
quote= 0;
}else if(lastchar=='\\')
escape= 1;
else if(lastchar=='"')
quote= '"';
else if(lastchar=='\'')
quote= '\'';
else if(lastchar=='{')
nesting++;
else if(lastchar=='}'){
if(nesting--==0){
fputc(lastchar,outfile);
fputc('\n',outfile);
state= RULES;
return;
}
}
fputc(lastchar,outfile);
}while(lastchar!='\n');

if(gettkn()<=0)
break;
if(lastchar=='\n')
savechar();
putback();
}
}

/*:41*//*42:*/
#line 662 "./yfmt.w"

void
proc_ctail()
{
return_line_info= 1;
copychars();
output_effective_line();
while(gettkn()> 0)
copychars();
}


/*:42*/
#line 214 "./yfmt.w"


/*:18*//*19:*/
#line 217 "./yfmt.w"

Procfun process[]= {proc_limbo,proc_chead,proc_defns,proc_rules,
proc_cbody,proc_ctail};

/*:19*//*20:*/
#line 222 "./yfmt.w"

int
convertor()
{
while(token!=EOF)
process[state]();
return 0;
}

/*:20*//*39:*/
#line 519 "./yfmt.w"

int
ctangle_comment()
{
if(saveindex> 3&&memcmp(savebuffer,"/*",2)==0){
int i;

if(savebuffer[2]==':'){
for(i= 3;i<saveindex&&savebuffer[i]!='*';i++)
if(!isdigit(savebuffer[i]))
return 0;

return i+1<saveindex&&savebuffer[i+1]=='/';
}else if(isdigit(savebuffer[2])){
for(i= 3;i<saveindex&&savebuffer[i]!=':';i++)
if(!isdigit(savebuffer[i]))
return 0;

return saveindex-i> 3&&memcmp(savebuffer+i,":*/",3)==0;
}
}
return 0;
}

/*:39*//*43:*/
#line 675 "./yfmt.w"

int
main(int argc,char**argv)
{
int c;
char*outname= "-";

/*12:*/
#line 123 "./yfmt.w"

progname= strrchr(argv[0],'/');
if(progname)
progname++;
else
progname= argv[0];

/*:12*/
#line 682 "./yfmt.w"

while((c= getopt(argc,argv,"o:h"))!=EOF){
switch(c){
case'o':
outname= optarg;
break;
case'h':
printf("usage: yfmt [-o outfile][-h] filename\n");
exit(0);
default:
exit(1);
}
}

if(optind==argc)
usage_error("Not enough arguments.");
else if(argc-optind> 1)
usage_error("Too many arguments.");

/*7:*/
#line 84 "./yfmt.w"

if(strcmp(argv[optind],"-")==0)
infile= stdin;
else if((infile= fopen(argv[optind],"r"))==NULL){
error("Cannot open input file `%s': %s",argv[optind],strerror(errno));
exit(1);
}

/*:7*/
#line 701 "./yfmt.w"

/*8:*/
#line 93 "./yfmt.w"

if(strcmp(outname,"-")==0)
outfile= stdout;
else if((outfile= fopen(outname,"w"))==NULL){
error("Cannot open output file `%s': %s",outname,strerror(errno));
exit(1);
}

/*:8*/
#line 702 "./yfmt.w"


/*5:*/
#line 75 "./yfmt.w"

real_locus.file= effective_locus.file= argv[optind];
real_locus.line= effective_locus.line= 1;

/*:5*/
#line 704 "./yfmt.w"


exit(convertor());
}


/*:43*/
